### PR TITLE
build-sys: peek golangci-lint and revive version based on the Go version

### PR DESCRIPTION
**Pull request by amery**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new script to dynamically retrieve and compare Go language versions for improved version management.
	- Updated versioning for linting tools to automatically fetch the latest compatible versions.

- **Bug Fixes**
	- Enhanced version retrieval process to ensure accurate versioning of tools.

### CODE REVIEW

**Comment by amery** [comment]
> Here's the code health analysis summary for commits `8afe07e..634ea68`. [View details on DeepSource](https://app.deepsource.com/gh/darvaza-proxy/core/run/e9e487ec-0d65-4277-9366-ba9c395a6109/) ↗.
> 
> ### Analysis Summary
> Analyzer	Status	Summary	Link
> DeepSource Shell Logo**Shell**	❌ Failure	
> ❗ 4 occurences introduced
> [View Check](https://app.deepsource.com/gh/darvaza-proxy/core/run/e9e487ec-0d65-4277-9366-ba9c395a6109/shell/) ↗
> DeepSource Go Logo**Go**	✅ Success		[View Check](https://app.deepsource.com/gh/darvaza-proxy/core/run/e9e487ec-0d65-4277-9366-ba9c395a6109/go/) ↗
> > 💡 If you’re a repository administrator, you can configure the quality gates from the [settings](https://app.deepsource.com/gh/darvaza-proxy/core/settings/reporting).

@deepsource-io those variables can only hold numbers, as by `xy_ver()`.

**Comment by amery** [comment]
@karasz is this implementation better?

File: internal/build/get_version.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+xy_ver() {
+	local ver="${1:-}"
+	local x y
+
+	if echo "v${ver}" | grep -qe '^v[0-9]\+\.[0-9]\+\(\..*\|$\)'; then
+		x="$(echo "$ver" | cut -d. -f1)"
+		y="$(echo "$ver" | cut -d. -f2)"
+		echo $((x * 1000 + y))
+	fi
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply internal/build/get_version.sh]: I'm yet to see a sh implementation not supporting `local` or `-a` as `test` option. I do agree with the error handling. PR amended accordingly

File: Makefile [hunk 2]
```diff
@@ -10,8 +10,10 @@ TOOLSDIR := $(CURDIR)/internal/build
 TMPDIR ?= $(CURDIR)/.tmp
 OUTDIR ?= $(TMPDIR)
 
-GOLANGCI_LINT_VERSION ?= v1.55
-REVIVE_VERSION ?= v1.3.7
+# Dynamic version selection based on Go version
+# Format: $(TOOLSDIR)/get_version.sh <go_version> <tool_version1> <tool_version2> ..
+GOLANGCI_LINT_VERSION ?= $(shell $(TOOLSDIR)/get_version.sh 1.20 v1.55 v1.59)
```
Comment (karasz) [inline Makefile:15]: any particular reason for which we hold back on the lint version? Why `v1.59` when latest is `v1.61.0` ?


[tags](https://github.com/golangci/golangci-lint/tags)

**Review by karasz — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (amery) [reply Makefile:15]: > any particular reason for which we hold back on the lint version? Why `v1.59` when latest is `v1.61.0` ?
> 
> [tags](https://github.com/golangci/golangci-lint/tags)

in this branch go1.22 is the latest supported Go. and newer versions of golangci-lint include new checks that require deep changes to the code base, which is out of the scope of this PR

**Review by karasz — APPROVED** [review]

**Pull request merged**